src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -1358,6 +1361,22 @@ private void saveSecondaryInterfaces(final List<String> newCmdComps, final List<
 		}		
 	}
 
+	private CompletionStage<SQLResult<Void>> clientHandshakeAsync(final String userid, final String pwd, final String db, final boolean shouldRequestVersion) {
+		CompletionStage<Void> clientHandshakeCompletionStage = CompletableFuture.runAsync(() -> {
+			try { clientHandshake(user, pwd, database, shouldRequestVersion); }
+			catch (Throwable t) {
+				LOGGER.log(Level.WARNING, String.format("Handshake exception %s occurred in reconnect() with message %s", t.toString()));
+
+				// wrap the throwable in a completion exception so it can be handled by the caller
+				throw new CompletionException(t);
+			}
+		});
+
+		return clientHandshakeCompletionStage.thenCompose(v -> clientHandshakeFinalize());
```

Comment:
If `clientHandshake(...)` completes exceptionally, I think the returned `CompletionStage<SQLResult<Void>>` will also complete exceptionally, which we actually don't want here. We want to propagate the exception through the `SQLResult` container.
```suggestion
                return runAsync_FIXME(() -> clientHandshake(user, pwd, database, shouldRequestVersion))
                    .thenCompose(successOrEx -> successOrEx
                        .ampSuccessAsync(ign -> clientHandshakeFinalize()));
```